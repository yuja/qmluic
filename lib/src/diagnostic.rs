//! Utility for error reporting.

use crate::qmlast::ParseError;
use camino::{Utf8Path, Utf8PathBuf};
use std::ops::Range;
use std::slice;

/// Type (or level) of diagnostic message.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum DiagnosticKind {
    Error,
}

/// Diagnostic message.
#[derive(Clone, Debug)]
pub struct Diagnostic {
    kind: DiagnosticKind,
    byte_range: Range<usize>,
    message: String,
    labels: Vec<(Range<usize>, String)>,
}

impl Diagnostic {
    /// Creates new diagnostic message of the given `kind`.
    pub fn new<S>(kind: DiagnosticKind, byte_range: Range<usize>, message: S) -> Self
    where
        S: Into<String>,
    {
        Diagnostic {
            kind,
            byte_range,
            message: message.into(),
            labels: Vec::new(),
        }
    }

    /// Creates new error message.
    pub fn error<S>(byte_range: Range<usize>, message: S) -> Self
    where
        S: Into<String>,
    {
        Self::new(DiagnosticKind::Error, byte_range, message)
    }

    /// Builds diagnostic with the given label attached.
    pub fn with_label<S>(mut self, byte_range: Range<usize>, message: S) -> Self
    where
        S: Into<String>,
    {
        self.push_label(byte_range, message);
        self
    }

    /// Builds diagnostic with the given labels attached.
    pub fn with_labels<I, S>(mut self, labels: I) -> Self
    where
        I: IntoIterator<Item = (Range<usize>, S)>,
        S: Into<String>,
    {
        self.extend_labels(labels);
        self
    }

    pub fn kind(&self) -> DiagnosticKind {
        self.kind
    }

    pub fn start_byte(&self) -> usize {
        self.byte_range.start
    }

    pub fn end_byte(&self) -> usize {
        self.byte_range.end
    }

    pub fn byte_range(&self) -> Range<usize> {
        Range {
            start: self.byte_range.start,
            end: self.byte_range.end,
        }
    }

    pub fn message(&self) -> &str {
        &self.message
    }

    pub fn labels(&self) -> &[(Range<usize>, String)] {
        &self.labels
    }

    /// Add a label to this diagnostic.
    pub fn push_label<S>(&mut self, byte_range: Range<usize>, message: S)
    where
        S: Into<String>,
    {
        self.labels.push((byte_range, message.into()));
    }

    /// Add labels to this diagnostic.
    pub fn extend_labels<I, S>(&mut self, labels: I)
    where
        I: IntoIterator<Item = (Range<usize>, S)>,
        S: Into<String>,
    {
        self.labels
            .extend(labels.into_iter().map(|(r, s)| (r, s.into())));
    }
}

impl From<&ParseError<'_>> for Diagnostic {
    fn from(error: &ParseError) -> Self {
        Self::error(error.byte_range(), error.to_string())
    }
}

impl From<ParseError<'_>> for Diagnostic {
    fn from(error: ParseError) -> Self {
        Self::from(&error)
    }
}

/// Manages diagnostic messages.
#[derive(Clone, Debug, Default)]
pub struct Diagnostics {
    diagnostics: Vec<Diagnostic>,
}

impl Diagnostics {
    pub fn new() -> Self {
        Diagnostics::default()
    }

    pub fn is_empty(&self) -> bool {
        self.diagnostics.is_empty()
    }

    pub fn len(&self) -> usize {
        self.diagnostics.len()
    }

    pub fn iter(&self) -> slice::Iter<'_, Diagnostic> {
        self.diagnostics.iter()
    }

    pub fn push<T>(&mut self, diag: T)
    where
        T: Into<Diagnostic>,
    {
        self.diagnostics.push(diag.into())
    }

    /// Extracts error from the given `result` and pushes it. Returns the success value if any.
    pub fn consume_err<T, E>(&mut self, result: Result<T, E>) -> Option<T>
    where
        E: Into<Diagnostic>,
    {
        match result {
            Ok(x) => Some(x),
            Err(e) => {
                self.push(e);
                None
            }
        }
    }
}

impl<'a> IntoIterator for &'a Diagnostics {
    type Item = &'a Diagnostic;
    type IntoIter = slice::Iter<'a, Diagnostic>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

impl<T> Extend<T> for Diagnostics
where
    T: Into<Diagnostic>,
{
    fn extend<I>(&mut self, iter: I)
    where
        I: IntoIterator<Item = T>,
    {
        self.diagnostics.extend(iter.into_iter().map(|d| d.into()))
    }
}

/// Manages per-file store of diagnostic messages.
#[derive(Clone, Debug, Default)]
pub struct ProjectDiagnostics {
    file_diagnostics: Vec<(Utf8PathBuf, Diagnostics)>, // in insertion order
}

impl ProjectDiagnostics {
    pub fn new() -> Self {
        ProjectDiagnostics::default()
    }

    pub fn is_empty(&self) -> bool {
        // push() guarantees that each file store is not empty
        self.file_diagnostics.is_empty()
    }

    pub fn iter(&self) -> slice::Iter<'_, (Utf8PathBuf, Diagnostics)> {
        self.file_diagnostics.iter()
    }

    pub fn diagnostics(&self) -> impl Iterator<Item = (&Utf8Path, &Diagnostic)> {
        self.file_diagnostics
            .iter()
            .flat_map(|(p, ds)| ds.iter().map(|d| (p.as_ref(), d)))
    }

    pub fn push<P>(&mut self, path: P, diagnostics: Diagnostics)
    where
        P: AsRef<Utf8Path>,
    {
        if !diagnostics.is_empty() {
            self.file_diagnostics
                .push((path.as_ref().to_owned(), diagnostics));
        }
    }
}
