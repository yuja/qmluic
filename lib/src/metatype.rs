//! Qt metatypes.json data types.

use serde::{Deserialize, Serialize};

// TODO: Symbols are stored as owned strings, but we'll eventually want to intern them,
// where we can probably remove these temporary owned strings at all.

/// C++ access specifier.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum AccessSpecifier {
    Private,
    Protected,
    Public,
}

/// Revision number of method or property.
#[derive(
    Clone, Copy, Debug, Default, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize,
)]
#[repr(transparent)]
#[serde(transparent)]
pub struct Revision(pub u32);

impl Revision {
    pub fn is_zero(&self) -> bool {
        self.0 == 0
    }
}

/// Class metadata.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Class {
    pub class_name: String,
    pub qualified_class_name: String,
    #[serde(default)]
    pub super_classes: Vec<SuperClassSpecifier>,
    #[serde(default)]
    pub class_infos: Vec<ClassInfo>,
    #[serde(default)]
    pub interfaces: Vec<Vec<Interface>>, // no idea why this is nested array
    #[serde(default)]
    pub object: bool,
    #[serde(default)]
    pub gadget: bool,
    #[serde(default)]
    pub namespace: bool,

    #[serde(default)]
    pub enums: Vec<Enum>,
    #[serde(default)]
    pub properties: Vec<Property>,
    #[serde(default)]
    pub constructors: Vec<Method>, // only QObject appears to define this
    #[serde(default)]
    pub signals: Vec<Method>,
    #[serde(default)]
    pub slots: Vec<Method>,
    #[serde(default)]
    pub methods: Vec<Method>,
}

impl Class {
    /// Creates class metadata of object type.
    pub fn new<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        let name: String = name.into();
        Class {
            class_name: unqualify_name(&name).to_owned(),
            qualified_class_name: name,
            object: true,
            ..Default::default()
        }
    }

    /// Creates class metadata of gadget type.
    pub fn new_gadget<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        let name: String = name.into();
        Class {
            class_name: unqualify_name(&name).to_owned(),
            qualified_class_name: name,
            gadget: true,
            ..Default::default()
        }
    }

    /// Creates class metadata of object type with public super classes.
    pub fn with_supers<S, I>(name: S, supers: I) -> Self
    where
        S: Into<String>,
        I: IntoIterator,
        I::Item: Into<String>,
    {
        let name: String = name.into();
        let super_classes = supers
            .into_iter()
            .map(SuperClassSpecifier::public)
            .collect();
        Class {
            class_name: unqualify_name(&name).to_owned(),
            qualified_class_name: name,
            object: true,
            super_classes,
            ..Default::default()
        }
    }
}

fn unqualify_name(name: &str) -> &str {
    name.rsplit_once("::").map(|(_, n)| n).unwrap_or(name)
}

/// Super class reference with the access specifier.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SuperClassSpecifier {
    pub name: String,
    pub access: AccessSpecifier,
}

impl SuperClassSpecifier {
    /// Creates public super class specifier.
    pub fn public<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        SuperClassSpecifier {
            name: name.into(),
            access: AccessSpecifier::Public,
        }
    }
}

/// Extra class metadata. (e.g. `""QML.Element"`" name)
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClassInfo {
    pub name: String,
    pub value: String,
}

impl ClassInfo {
    /// Creates class info pair.
    pub fn new<S, T>(name: S, value: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        ClassInfo {
            name: name.into(),
            value: value.into(),
        }
    }
}

/// Qt plugin interface identifier. (see `Q_DECLARE_INTERFACE()`)
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Interface {
    pub class_name: String,
    pub id: String,
}

/// Enum (and flag) metadata.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Enum {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alias: Option<String>,
    pub is_class: bool,
    pub is_flag: bool,
    pub values: Vec<String>,
}

impl Enum {
    /// Creates enum metadata.
    pub fn new<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        Enum {
            name: name.into(),
            ..Default::default()
        }
    }

    /// Creates flag metadata.
    pub fn new_flag<S, T>(name: S, alias: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        Enum {
            name: name.into(),
            alias: Some(alias.into()),
            is_flag: true,
            ..Default::default()
        }
    }

    /// Creates enum metadata with values.
    pub fn with_values<S, I>(name: S, values: I) -> Self
    where
        S: Into<String>,
        I: IntoIterator,
        I::Item: Into<String>,
    {
        Enum {
            name: name.into(),
            values: values.into_iter().map(|n| n.into()).collect(),
            ..Default::default()
        }
    }
}

/// Property metadata.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Property {
    pub name: String,
    pub r#type: String,

    /// See `Q_PRIVATE_PROPERTY()`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub private_class: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub member: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub read: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub write: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reset: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub notify: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bindable: Option<String>,

    #[serde(default, skip_serializing_if = "Revision::is_zero")]
    pub revision: Revision,
    pub designable: bool,
    pub scriptable: bool,
    pub stored: bool,
    pub user: bool,
    pub constant: bool,
    pub r#final: bool,
    pub required: bool,

    /// Property index in the current meta object. (new in Qt 6)
    pub index: Option<i32>,
}

impl Property {
    /// Creates property metadata.
    pub fn new<S, T>(name: S, type_name: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        Property {
            name: name.into(),
            r#type: type_name.into(),
            ..Default::default()
        }
    }

    /// Creates property metadata of final.
    pub fn new_final<S, T>(name: S, type_name: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        Property {
            name: name.into(),
            r#type: type_name.into(),
            r#final: true,
            ..Default::default()
        }
    }
}

impl Default for Property {
    fn default() -> Self {
        Self {
            name: String::new(),
            r#type: String::new(),
            private_class: None,
            member: None,
            read: None,
            write: None,
            reset: None,
            notify: None,
            bindable: None,
            revision: Revision(0),
            designable: true,
            scriptable: true,
            stored: true,
            user: false,
            constant: false,
            r#final: false,
            required: false,
            index: None,
        }
    }
}

/// Signal, slot, or callable function metadata.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Method {
    pub name: String,
    pub access: AccessSpecifier,
    pub return_type: String,
    #[serde(default)]
    pub arguments: Vec<Argument>,
    #[serde(default, skip_serializing_if = "Revision::is_zero")]
    pub revision: Revision,
}

impl Default for Method {
    fn default() -> Self {
        Self {
            name: String::new(),
            access: AccessSpecifier::Public,
            return_type: String::new(),
            arguments: vec![],
            revision: Revision(0),
        }
    }
}

/// Signal, slot, or callable function argument.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Argument {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub r#type: String,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CompilationUnit {
    #[serde(default)]
    pub classes: Vec<Class>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_file: Option<String>,
    #[serde(default)]
    pub output_revision: i32,
}

/// Collects all classes from metatypes.json document.
///
/// The JSON document should consist of either a compilation unit object `{"classes": ...}`
/// or an array of such objects `[{"classes": ...}, ...]`.
pub fn extract_classes_from_str(json_data: &str) -> serde_json::Result<Vec<Class>> {
    // Don't use untagged enum to dispatch [{...}] and {...}. Doing that would move any
    // parsing error to the document level and line/column information would be lost.
    if json_data.trim_start().starts_with('{') {
        let unit: CompilationUnit = serde_json::from_str(json_data)?;
        Ok(unit.classes)
    } else {
        let units: Vec<CompilationUnit> = serde_json::from_str(json_data)?;
        Ok(units.into_iter().flat_map(|u| u.classes).collect())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize_class_decl() {
        let data = r###"{
            "className": "QAction",
            "enums": [
                {
                    "isClass": false,
                    "isFlag": false,
                    "name": "Priority",
                    "values": [
                        "LowPriority",
                        "NormalPriority",
                        "HighPriority"
                    ]
                }
            ],
            "object": true,
            "properties": [
                {
                    "constant": false,
                    "designable": true,
                    "final": false,
                    "name": "checkable",
                    "notify": "changed",
                    "read": "isCheckable",
                    "required": false,
                    "scriptable": true,
                    "stored": true,
                    "type": "bool",
                    "user": false,
                    "write": "setCheckable"
                }
            ],
            "qualifiedClassName": "QAction",
            "signals": [
                {
                    "access": "public",
                    "name": "changed",
                    "returnType": "void"
                },
                {
                    "access": "public",
                    "arguments": [
                        {
                            "name": "checked",
                            "type": "bool"
                        }
                    ],
                    "name": "triggered",
                    "returnType": "void"
                }
            ],
            "slots": [
                {
                    "access": "public",
                    "name": "trigger",
                    "returnType": "void"
                },
                {
                    "access": "public",
                    "name": "hover",
                    "returnType": "void"
                },
                {
                    "access": "public",
                    "arguments": [
                        {
                            "type": "bool"
                        }
                    ],
                    "name": "setChecked",
                    "returnType": "void"
                }
            ],
            "superClasses": [
                {
                    "access": "public",
                    "name": "QObject"
                }
            ]
        }"###;
        let _meta: Class = serde_json::from_str(data).unwrap();
    }

    #[test]
    fn extract_classes_unit() {
        let data = r###"{
            "classes": [
                {"className": "Foo", "qualifiedClassName": "Foo", "object": true},
                {"className": "Bar", "qualifiedClassName": "Bar", "object": true}
            ]
        }"###;
        let classes = extract_classes_from_str(data).unwrap();
        assert_eq!(classes.len(), 2);
    }

    #[test]
    fn extract_classes_nested() {
        let data = r###"[
            {"classes": [{"className": "Foo", "qualifiedClassName": "Foo", "object": true}]},
            {"classes": [{"className": "Bar", "qualifiedClassName": "Bar", "object": true}]}
        ]"###;
        let classes = extract_classes_from_str(data).unwrap();
        assert_eq!(classes.len(), 2);
    }
}
