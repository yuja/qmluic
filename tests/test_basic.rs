use qmluic::diagnostic::Diagnostics;
use qmluic::metatype;
use qmluic::qmlast::UiDocument;
use qmluic::typemap::TypeMap;
use qmluic::uigen::XmlWriter;
use qmluic_cli::UiBuilder;
use std::fs;
use std::path::Path;

#[test]
fn test_translate_example() {
    insta::assert_snapshot!(translate_file("examples/SettingsDialog.qml",));
    insta::assert_snapshot!(translate_file("examples/MainWindow.qml"));
    insta::assert_snapshot!(translate_file("examples/VariousLayouts.qml"));
}

fn translate_file(path: impl AsRef<Path>) -> String {
    let mut type_map = TypeMap::with_primitive_types();
    type_map.extend(load_metatypes());
    let doc = UiDocument::read(path).unwrap();
    let mut diagnostics = Diagnostics::new();
    let form = UiBuilder::new(&type_map, &doc, &mut diagnostics)
        .build()
        .unwrap();
    let mut buf = Vec::new();
    form.serialize_to_xml(&mut XmlWriter::new_with_indent(&mut buf, b' ', 1))
        .unwrap();
    String::from_utf8(buf).unwrap()
}

fn load_metatypes() -> Vec<metatype::Class> {
    let paths = [
        "contrib/metatypes/qmluic_metatypes.json",
        "contrib/metatypes/qt5core_metatypes.json",
        "contrib/metatypes/qt5gui_metatypes.json",
        "contrib/metatypes/qt5widgets_metatypes.json",
    ];
    paths
        .iter()
        .flat_map(|p| {
            let data = fs::read_to_string(p).unwrap();
            metatype::extract_classes_from_str(&data).unwrap()
        })
        .collect()
}
