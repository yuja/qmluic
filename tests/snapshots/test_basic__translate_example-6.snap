---
source: tests/test_basic.rs
expression: "common::translate_file(\"examples/thg/HgEmailDialog.qml\").unwrap()"
---
<ui version="4.0">
 <class>HgEmailDialog</class>
 <widget class="QDialog">
  <property name="height">
   <number>520</number>
  </property>
  <property name="sizeGripEnabled">
   <bool>true</bool>
  </property>
  <property name="width">
   <number>660</number>
  </property>
  <property name="windowTitle">
   <string>Email</string>
  </property>
  <layout class="QVBoxLayout">
   <item>
    <widget class="QTabWidget" name="mainTabs">
     <property name="currentIndex">
      <number>0</number>
     </property>
     <property name="documentMode">
      <bool>false</bool>
     </property>
     <property name="movable">
      <bool>false</bool>
     </property>
     <property name="tabsClosable">
      <bool>false</bool>
     </property>
     <widget class="QWidget" name="editTab">
      <attribute name="title">
       <string>Edit</string>
      </attribute>
      <layout class="QGridLayout">
       <item column="0" row="0">
        <widget class="QGroupBox" name="envelopeBox">
         <layout class="QFormLayout">
          <property name="fieldGrowthPolicy">
           <enum>QFormLayout::ExpandingFieldsGrow</enum>
          </property>
          <item column="0" row="0">
           <widget class="QLabel">
            <property name="buddy">
             <cstring>toEdit</cstring>
            </property>
            <property name="text">
             <string>To:</string>
            </property>
           </widget>
          </item>
          <item column="1" row="0">
           <widget class="QComboBox" name="toEdit">
            <property name="editable">
             <bool>true</bool>
            </property>
            <property name="insertPolicy">
             <enum>QComboBox::InsertAtTop</enum>
            </property>
            <property name="sizePolicy">
             <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
             </sizepolicy>
            </property>
           </widget>
          </item>
          <item column="0" row="1">
           <widget class="QLabel">
            <property name="buddy">
             <cstring>ccEdit</cstring>
            </property>
            <property name="text">
             <string>Cc:</string>
            </property>
           </widget>
          </item>
          <item column="1" row="1">
           <widget class="QComboBox" name="ccEdit">
            <property name="editable">
             <bool>true</bool>
            </property>
            <property name="insertPolicy">
             <enum>QComboBox::InsertAtTop</enum>
            </property>
            <property name="sizePolicy">
             <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
             </sizepolicy>
            </property>
           </widget>
          </item>
          <item column="0" row="2">
           <widget class="QLabel">
            <property name="buddy">
             <cstring>fromEdit</cstring>
            </property>
            <property name="text">
             <string>From:</string>
            </property>
           </widget>
          </item>
          <item column="1" row="2">
           <widget class="QComboBox" name="fromEdit">
            <property name="editable">
             <bool>true</bool>
            </property>
            <property name="insertPolicy">
             <enum>QComboBox::InsertAtTop</enum>
            </property>
            <property name="sizePolicy">
             <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
             </sizepolicy>
            </property>
           </widget>
          </item>
          <item column="0" row="3">
           <widget class="QLabel">
            <property name="buddy">
             <cstring>inReplyToEdit</cstring>
            </property>
            <property name="text">
             <string>In-Reply-To:</string>
            </property>
           </widget>
          </item>
          <item column="1" row="3">
           <widget class="QLineEdit" name="inReplyToEdit">
            <property name="toolTip">
             <string>Message identifier to reply to, for threading</string>
            </property>
           </widget>
          </item>
          <item column="0" row="4">
           <widget class="QLabel">
            <property name="buddy">
             <cstring>flagEdit</cstring>
            </property>
            <property name="text">
             <string>Flag:</string>
            </property>
           </widget>
          </item>
          <item column="1" row="4">
           <widget class="QComboBox" name="flagEdit">
            <property name="editable">
             <bool>true</bool>
            </property>
            <property name="insertPolicy">
             <enum>QComboBox::InsertAtTop</enum>
            </property>
            <property name="sizePolicy">
             <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
             </sizepolicy>
            </property>
           </widget>
          </item>
         </layout>
        </widget>
       </item>
       <item column="1" row="0">
        <widget class="QGroupBox" name="optionsEdit">
         <property name="sizePolicy">
          <sizepolicy hsizetype="Maximum" vsizetype="Preferred">
          </sizepolicy>
         </property>
         <layout class="QVBoxLayout">
          <item>
           <widget class="QFrame" name="patchFrame">
            <property name="frameShadow">
             <enum>QFrame::Raised</enum>
            </property>
            <property name="frameShape">
             <enum>QFrame::NoFrame</enum>
            </property>
            <layout class="QVBoxLayout">
             <item>
              <widget class="QRadioButton" name="hgPatchRadio">
               <property name="text">
                <string>Send changesets as Hg patches</string>
               </property>
               <property name="whatsThis">
                <string>Hg patches (as generated by export command) are compatible with most patch programs. They include a header which contains the most important changeset metadata.</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QRadioButton" name="gitPatchRadio">
               <property name="text">
                <string>Use extended (git) patch format</string>
               </property>
               <property name="whatsThis">
                <string>Git patches can describe binary files, copies, and permission changes, but recipients may not be able to use them if they are not using git or Mercurial.</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QRadioButton" name="plainPatchRadio">
               <property name="text">
                <string>Plain, do not prepend Hg header</string>
               </property>
               <property name="whatsThis">
                <string>Stripping Mercurial header removes username and parent information.  Only useful if recipient is not using Mercurial (and does not like to see the headers).</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QRadioButton" name="bundleRadio">
               <property name="text">
                <string>Send single binary bundle, not patches</string>
               </property>
               <property name="whatsThis">
                <string>Bundles store complete changesets in binary form. Upstream users can pull from them. This is the safest way to send changes to recipient Mercurial users.</string>
               </property>
              </widget>
             </item>
            </layout>
           </widget>
          </item>
          <item>
           <widget class="QFrame" name="extraFrame">
            <property name="frameShadow">
             <enum>QFrame::Raised</enum>
            </property>
            <property name="frameShape">
             <enum>QFrame::NoFrame</enum>
            </property>
            <layout class="QHBoxLayout">
             <item>
              <widget class="QCheckBox" name="bodyCheck">
               <property name="checked">
                <bool>true</bool>
               </property>
               <property name="text">
                <string>body</string>
               </property>
               <property name="toolTip">
                <string>Send patches as part of the email body</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QCheckBox" name="attachCheck">
               <property name="text">
                <string>attach</string>
               </property>
               <property name="toolTip">
                <string>Send patches as attachments</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QCheckBox" name="inlineCheck">
               <property name="text">
                <string>inline</string>
               </property>
               <property name="toolTip">
                <string>Send patches as inline attachments</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QCheckBox" name="diffstatCheck">
               <property name="text">
                <string>diffstat</string>
               </property>
               <property name="toolTip">
                <string>Add diffstat output to messages</string>
               </property>
              </widget>
             </item>
             <item>
              <spacer>
               <property name="orientation">
                <enum>Qt::Horizontal</enum>
               </property>
              </spacer>
             </item>
            </layout>
           </widget>
          </item>
         </layout>
        </widget>
       </item>
       <item column="0" colspan="2" row="1">
        <widget class="QCheckBox" name="writeIntroCheck">
         <property name="text">
          <string>Write patch series (bundle) description</string>
         </property>
         <property name="whatsThis">
          <string>Patch series description is sent in initial summary email with [PATCH 0 of N] subject.  It should describe the effects of the entire patch series.  When emailing a bundle, these fields make up the message subject and body. Flags is a comma separated list of tags which are inserted into the message subject prefix.</string>
         </property>
        </widget>
       </item>
       <item column="0" colspan="2" row="2">
        <widget class="QSplitter">
         <property name="orientation">
          <enum>Qt::Vertical</enum>
         </property>
         <widget class="QGroupBox" name="introBox">
          <layout class="QVBoxLayout">
           <item>
            <layout class="QHBoxLayout">
             <item>
              <widget class="QLabel">
               <property name="text">
                <string>Subject:</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QComboBox" name="subjectEdit">
               <property name="editable">
                <bool>true</bool>
               </property>
               <property name="insertPolicy">
                <enum>QComboBox::InsertAtTop</enum>
               </property>
               <property name="sizePolicy">
                <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
                </sizepolicy>
               </property>
              </widget>
             </item>
            </layout>
           </item>
           <item>
            <widget class="QPlainTextEdit">
             <property name="font">
              <font>
               <family notr="true">Monospace</family>
              </font>
             </property>
            </widget>
           </item>
          </layout>
         </widget>
         <widget class="QGroupBox" name="changesetsBox">
          <property name="title">
           <string>Changesets</string>
          </property>
          <layout class="QVBoxLayout">
           <item>
            <widget class="QTreeView" name="changesetsView">
             <property name="indentation">
              <number>0</number>
             </property>
             <property name="itemsExpandable">
              <bool>false</bool>
             </property>
             <property name="rootIsDecorated">
              <bool>false</bool>
             </property>
            </widget>
           </item>
           <item>
            <layout class="QHBoxLayout">
             <item>
              <widget class="QPushButton" name="selectAllButton">
               <property name="text">
                <string>Select &amp;All</string>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QPushButton" name="selectNoneButton">
               <property name="text">
                <string>Select &amp;None</string>
               </property>
              </widget>
             </item>
             <item>
              <spacer>
               <property name="orientation">
                <enum>Qt::Horizontal</enum>
               </property>
              </spacer>
             </item>
            </layout>
           </item>
          </layout>
         </widget>
        </widget>
       </item>
      </layout>
     </widget>
     <widget class="QWidget" name="previewTab">
      <attribute name="title">
       <string>Preview</string>
      </attribute>
      <layout class="QGridLayout">
       <item column="0" row="0">
        <widget class="QPlainTextEdit" name="previewEdit">
        </widget>
       </item>
      </layout>
     </widget>
    </widget>
   </item>
   <item>
    <layout class="QHBoxLayout">
     <item>
      <widget class="QPushButton" name="settingButton">
       <property name="default">
        <bool>false</bool>
       </property>
       <property name="text">
        <string>&amp;Settings</string>
       </property>
       <property name="toolTip">
        <string>Configure email settings</string>
       </property>
      </widget>
     </item>
     <item>
      <spacer>
       <property name="orientation">
        <enum>Qt::Horizontal</enum>
       </property>
      </spacer>
     </item>
     <item>
      <widget class="QPushButton" name="sendButton">
       <property name="default">
        <bool>false</bool>
       </property>
       <property name="enabled">
        <bool>false</bool>
       </property>
       <property name="text">
        <string>Send &amp;Email</string>
       </property>
      </widget>
     </item>
     <item>
      <widget class="QPushButton" name="closeButton">
       <property name="default">
        <bool>true</bool>
       </property>
       <property name="text">
        <string>&amp;Close</string>
       </property>
      </widget>
     </item>
    </layout>
   </item>
  </layout>
 </widget>
</ui>

